### PR TITLE
[6.x] PHP 7.4 support

### PR DESCRIPTION
**Pull request by driesvints**
This fixes our own PHP 7.4 build failures. PHP 7.4 is out in November so it's best that we make use of this major release to do the necessarily breaking changes already.

There's two remaining failures:

- A problem with `ArrayInput` from Symfony which is already fixed (https://github.com/symfony/symfony/pull/32806). Build will be fixed once Symfony 4.3.4 is out.
- Unbinding `$this` of closure is deprecated within Carbon which @kylekatarnls will need to fix. I'd like to send in a PR but I think he will probably figure out the internals much faster than me.

Problems fixed with this PR:

- Fix array_key_exists call on object: https://github.com/laravel/framework/commit/9afccd60d5f28d8a792a2b6ca661441e943e1268
- Fix CreateMembersTable resolving: https://github.com/laravel/framework/commit/0a3563ef0301fc133bcce9458fb374ee6054a11b
- Fix array call on null for Facade (https://github.com/laravel/framework/commit/7743cba56ef53f5da8ee650064c18d8cafe55ff7)
- Fix unbinding `$this` for closures (https://github.com/laravel/framework/commit/c9f0144378130e0a3528d8b2b7513b16b450045e). Symfony fixed these problems as well here: https://github.com/symfony/symfony/pull/32796

The unbinding of closures is the most significant one. Basically no internals in Laravel were changed. In PHP 7.4 people will need to use the `static` keyword when they want to allow to statically call closure macros. But this is a language limitation and not a Laravel one.

Fixes https://github.com/laravel/framework/issues/29411

### CODE REVIEW

**Comment by derekmd** [comment]
Tests are green but `Request@validate()` and `hasValidSignature()` go through through a macro registered within the framework, which would now be broken.

https://github.com/laravel/framework/blob/c9f0144378130e0a3528d8b2b7513b16b450045e/src/Illuminate/Foundation/Providers/FoundationServiceProvider.php#L50-L66

`$this->all()` to fetch the payload can no longer be referenced in the `Closure`.

I mentioned this in the Discord #internals channel, but PHP 8.0 doing this is a massive breaking change to Laravel's `Macroable` ecosystem. Almost every package that uses the `Macroable` trait will no longer work.

----

### Collection macros are dead?

By only changing the `Closure` to be static, `Collection` macros would no longer be possible. Instead the `Closure` must be passed the instance being worked on.

This macro in Laravel 5.8:

```php
Collection::macro('takeWhile', function (callable $callback) {
    $items = [];

    foreach ($this->items as $key => $item) {
        if (!$callback($item, $key)) {
            break;
        }

        $items[$key] = $item;
    }

    return new static($items);
});
```

in Laravel 6.0 would have to be:

```php
Collection::macro('takeWhile', function ($collection, callable $callback) {
    $items = [];

    foreach ($collection->all() as $key => $item) {
        if (!$callback($item, $key)) {
            break;
        }

        $items[$key] = $item;
    }

    return new static($items);
});
```

The only way that could work is to conditionally change the parameters passed into the closure depending on if you're in a `static` context.

```php
if ($macro instanceof Closure) {
    if ($this === null) {
        return call_user_func_array($macro->bindTo($this, static::class), $parameters);
    }

    return call_user_func_array($macro->bindTo($this, static::class), array_merge($this, $parameters));
}
```

**Comment by taylorotwell** [comment]
Can someone link to further explanation about this Closure breaking change? Will Closure macros even be possible anymore?

**Comment by thecrypticace** [comment]
They're completely possible and would still work as intended. That part of this PR doesn't seem correct. PHP 7.4 warns against the removal of $this from normal Closures but you can still change `$this` without problem.

See this for a more detailed explanation in code: https://3v4l.org/ofaq7

**Comment by driesvints** [comment]
I've pushed a new commit but it now throws a segmentation fault 🤔  https://travis-ci.org/driesvints/framework/jobs/570839855

File: src/Illuminate/Support/Traits/Macroable.php
Comment (kylekatarnls) [inline src/Illuminate/Support/Traits/Macroable.php]: Does this mean `static::` can no longer be used inside static macros?

File: tests/Support/SupportMacroableTest.php
Comment (kylekatarnls) [inline tests/Support/SupportMacroableTest.php]: Laravel 6 will lose this feature? Isn't a bit sad to disable the feature (rather than deprecate it) since it's just deprecated and not yet actually removed?

File: src/Illuminate/Support/Traits/Macroable.php
Comment (driesvints) [reply src/Illuminate/Support/Traits/Macroable.php]: If this change is merged then that won't be possible no.

File: tests/Support/SupportMacroableTest.php
Comment (driesvints) [reply tests/Support/SupportMacroableTest.php]: If it's deprecated in PHP then it'll be removed eventually. Why wait?

Comment (kylekatarnls) [reply tests/Support/SupportMacroableTest.php]: By deprecate it PHP warn before removing it. If you remove it now in Laravel there is no warning time for Laravel users to realize there macros no longer work. It's even worst for libraries that would have a range of compatibility like laravel "^5.7 || ^6.0"

Or the opposite case: if you have a Laravel 6.0 app with PHP 7.4, you actually could use `static::` if you did not plan to upgrade yet/soon to PHP 8, Laravel hurry you to refactor your macros.

It's OK to remove in 6.0, but it sounds better to deprecate in a version and remove in the next one. Maybe it could be deprecated in a patch of 5.8.

File: src/Illuminate/Support/Traits/Macroable.php
Comment (nikic) [reply src/Illuminate/Support/Traits/Macroable.php]: I have no familiarity with this macro concept, but I'm wondering if you could just rebind the scope here while leaving `$this` alone?

Comment (driesvints) [reply src/Illuminate/Support/Traits/Macroable.php]: @nikic hmm not sure what the method call for that would be. Afaik both `bind` and `bindTo` require `$newthis`.

Comment (nikic) [reply src/Illuminate/Support/Traits/Macroable.php]: It's a bit ugly, but you could try passing `(new ReflectionFunction($closure))->getClosureThis()`.

Comment (driesvints) [reply src/Illuminate/Support/Traits/Macroable.php]: Ah, I was actually thinking that reflection might help here. Thanks for suggesting that 👍 

I'll try to give it a go.

Comment (driesvints) [reply src/Illuminate/Support/Traits/Macroable.php]: Hmm seems that didn't produce the wanted result. It still throws an error when I pass the closure. It does work when I pass in `static::class` (the class it places the macro on) but then two other tests fail.

```
public static function __callStatic($method, $parameters)
    {
        if (! static::hasMacro($method)) {
            throw new BadMethodCallException(sprintf(
                'Method %s::%s does not exist.', static::class, $method
            ));
        }

        $macro = static::$macros[$method];

        if ($macro instanceof Closure) {
            $newThis = (new ReflectionFunction(static::class))->getClosureThis();

            return call_user_func_array($macro->bindTo($newThis, static::class), $parameters);
        }

        return call_user_func_array($macro, $parameters);
    }
```

**Comment by nikic** [comment]
Regarding the segfault:

```
==12670== Process terminating with default action of signal 6 (SIGABRT)
==12670==    at 0xA127E97: raise (raise.c:51)
==12670==    by 0xA129800: abort (abort.c:79)
==12670==    by 0xA119399: __assert_fail_base (assert.c:92)
==12670==    by 0xA119411: __assert_fail (assert.c:101)
==12670==    by 0x947D10: _zend_is_inconsistent (zend_hash.c:71)
==12670==    by 0x94CDE6: zend_array_destroy (zend_hash.c:1591)
==12670==    by 0x98519B: zend_object_std_dtor (zend_objects.c:53)
==12670==    by 0x96F521: zend_gc_collect_cycles (zend_gc.c:1565)
==12670==    by 0x96CD8C: gc_possible_root_when_full (zend_gc.c:586)
==12670==    by 0x96CF11: gc_possible_root (zend_gc.c:636)
==12670==    by 0x9981D6: zend_object_release (zend_objects_API.h:77)
==12670==    by 0x9A8D6A: ZEND_DO_FCALL_SPEC_RETVAL_USED_HANDLER (zend_vm_execute.h:1754)
```

This looks like fallout from some recent GC changes, I think I know what cause this...

**Comment by driesvints** [comment]
@nikic thanks for looking into that!

@derekmd this would be resolved with the latest commits, right?

**Comment by derekmd** [comment]
@driesvints Yup, leaving `Macroable@__call()` unchanged will allow existing Collection macros to behave as normal.

I've submitted https://github.com/laravel/framework/pull/29535 for `Request::validate()` test coverage so future breaking changes will be caught.

File: src/Illuminate/Support/Traits/Macroable.php
Comment (nikic) [reply src/Illuminate/Support/Traits/Macroable.php]: This worked for me: https://gist.github.com/nikic/a6e97696d1351261951edeac97ef211a At least I'm not getting any test failures that look related (there are some in Carbon, but those use different code).

**Comment by nikic** [comment]
The segfault should be fixed in the next nightly build.

**Comment by GrahamCampbell** [comment]
I wonder if we should also bring these fixes into Laravel 5.8? They don't seem too heavy.

**Comment by driesvints** [comment]
@GrahamCampbell don't think it's worth the effort. 5.8 won't be supported anymore after this month.

**Comment by X-Coder264** [comment]
Like @GrahamCampbell said, these fixes don't seem too heavy and I personally think that it's weird that the current latest Laravel framework version that you can get (which was released just a couple of months ago) won't support the next _minor_ PHP version which is due to be released in just two-three months.

**Comment by driesvints** [comment]
Okay. I'll backport these once this is merged. I'd like to keep this PR open for now to continue the discussion.

File: src/Illuminate/Support/Traits/Macroable.php
Comment (driesvints) [reply src/Illuminate/Support/Traits/Macroable.php]: @nikic I think that might be because the failing test was removed because at first we thought this couldn't be supported anymore. I've pushed a new commit with these changes and re-added the test which is now failing. (force pushed this, sorry)

https://github.com/laravel/framework/pull/29482/commits/f3b451201c54c1fe5a3a848434a92e9f444c2cb9

Failing build: https://travis-ci.org/laravel/framework/builds/571421063

File: tests/Support/SupportMacroableTest.php
Comment (driesvints) [reply tests/Support/SupportMacroableTest.php]: Re-added this.

**Comment by kylekatarnls** [comment]
What? Existing static macros will not be backward compatibles, and you will add this in a minor release of 5.8? Muting the error in 5.8 to silently support PHP 7.4 is OK, but this is a breaking change. For proof, you removed a unit tests. Unit tests are supposed to be what you can do and rely on it for the current version.

File: src/Illuminate/Support/Traits/Macroable.php
Comment (nikic) [reply src/Illuminate/Support/Traits/Macroable.php]: I see. The issue is that you're using `static::` rather than `self::`, and `static` is determined by `get_class($this)` if `$this` exists, with the bound scope not playing into it.

I think we need to relax this deprecation to only the cases where `$this` is actually referenced inside the closure.

**Comment by devcircus** [comment]
Looks like this targets 6.0

File: src/Illuminate/Http/Resources/DelegatesToResource.php
Comment (taylorotwell) [inline src/Illuminate/Http/Resources/DelegatesToResource.php]: isset and array_key_exists are not the same behavior?

Comment (KennedyTedesco) [reply src/Illuminate/Http/Resources/DelegatesToResource.php]: `array_key_exists` is also faster than isset. Ref: https://github.com/php/php-src/pull/3360

Comment (nikic) [reply src/Illuminate/Http/Resources/DelegatesToResource.php]: @taylorotwell The behavior is different, but isset() is "correct" where the contract of offsetExists() is concerned (it should return false if the value is null, otherwise isset() on the object will not work correctly). From that perspective I think this is a fairly safe change.

@KennedyTedesco No, array_key_exists() is not faster in the way it is used here.

File: src/Illuminate/Support/Traits/Macroable.php
Comment (nikic) [reply src/Illuminate/Support/Traits/Macroable.php]: I think you should just ignore this issue for now -- I plan to relax the deprecation on the PHP side, after which (I think) no changes in Laravel should be needed anymore. You are only trying to unbind `$this` in cases where it is not used inside the closure, right?

File: src/Illuminate/Http/Resources/DelegatesToResource.php
Comment (KennedyTedesco) [reply src/Illuminate/Http/Resources/DelegatesToResource.php]: @nikic Thanks for pointing this out. 👍

File: src/Illuminate/Support/Traits/Macroable.php
Comment (driesvints) [reply src/Illuminate/Support/Traits/Macroable.php]: @nikic the test still fails if I change `static::` to `self::` [on this line](https://github.com/laravel/framework/blob/e0f075918dae4ac5975d80ade0af01dfd454883b/tests/Support/SupportMacroableTest.php#L46) it still fails. But that shouldn't be necessary since we're trying to make sure `static::` calls in closures still work.

> You are only trying to unbind $this in cases where it is not used inside the closure, right?
 
Correct yes. When using `static::` it could technically still be used in a non-static context as well but when you're calling the macro statically then there shouldn't be a concept of `$this` inside the closure.

When would these changes be implemented? We're planning to release 6.0 end of this month or early september. Depending on this I'd send the rest of this PR to the 5.8 branch already. 

Thank you for all your work on this btw :)

**Comment by driesvints** [comment]
Hey @kylekatarnls please see my conversation with @nikic here: https://github.com/laravel/framework/pull/29482#discussion_r314035974

We're trying to see if we might loosen up the new implementation a bit.

File: src/Illuminate/Http/Resources/DelegatesToResource.php
Comment (driesvints) [reply src/Illuminate/Http/Resources/DelegatesToResource.php]: @taylorotwell I think @nikic explained this well :)

**Comment by nikic** [comment]
I've merged the change to the closure $this deprecation, so you might want to give this another try.

**Comment by GrahamCampbell** [comment]
@nikic How do you tell if `$this` is used? What if there is a `require` statement in a Closure body?

**Comment by nikic** [comment]
@GrahamCampbell Only syntactical uses of `$this` are relevant here.

**Comment by GrahamCampbell** [comment]
@nikic Does `$this` not get passed to required files?

Consider the following:

file.php:
```php
<?php

echo get_class($this);
```

run.php:
```php
<?php

class Foo
{
    public function bar()
    {
        $fn = function () { require 'file.php'; };
        $fn();
    }
}

(new Foo)->bar();
```

```bash
$ php run.php
```

The following code will write `Foo` to the screen.

**Comment by nikic** [comment]
@GrahamCampbell To provide some context: The reason we're interested in this deprecation is exclusively to enable some performance improvements in PHP 8 related to $this accesses, which is enabled by the removal of static calls to non-static methods. $this accesses can be split into two categories: Those where we know $this to be non-null and those where we don't. Method calls (will) fall into the former category. With this deprecation closures will also fall into the former category. $this accesses inside free-standing (e.g. required) code fall into the latter. The change we have in mind will make the former category faster and the latter slower, so we'd like closures to fall into the fast case, while we don't particularly care about uses of $this in required code. This is also why indirect uses of $this through required code in closures are not relevant here: While they do use $this, they are not directly part of the closure.

From a semantic perspective it is cleaner to treat these (direct $this access and required $this access) the same, and a reasonable person may argue that it would be better to force the use of `static` on these closures. But as it breaks an existing use-case and we don't actually need this from a technical perspective, this is the compromise...

**Comment by GrahamCampbell** [comment]
@nikic Ah, I see. Thanks for the detailed explanation.

**Comment by driesvints** [comment]
@nikic when I retrigger the build it seems that we get a new segmentation fault: https://travis-ci.org/laravel/framework/jobs/571421072

**Comment by driesvints** [comment]
@nikic when I revert to the behavior before this PR I also get the segfault.

**Comment by nikic** [comment]
Thanks for pointing this out! I've bisected the segfault to https://github.com/php/php-src/commit/60a7e60b61b8e4a3d455974c83f76a26546ce117, will investigate further.

**Comment by nikic** [comment]
@driesvints Segfault should be fixed with https://github.com/php/php-src/commit/589542f50cbaf2bfc88aea79092df9db54f75ac6.

**Comment by driesvints** [comment]
@nikic okay. I suspect it'll take a while before it's released? Because atm I still see the segfault happening.

**Comment by nikic** [comment]
@driesvints A new 7.4snapshot has been built ~2h ago (https://travis-ci.org/travis-ci/php-src-builder/builds/577246273), but your build still used yesterday's version. Probably it takes some time to propagate artifacts from AWS to GCE.

**Comment by GrahamCampbell** [comment]
Would it be too much to ask for this to be backported to L5.5 as well as 5.8? It looks like this PR is nice and small now. ;)

**Comment by driesvints** [comment]
@GrahamCampbell we first need to make sure the current build is passing before we even begin to backport this anywhere.

**Comment by kylekatarnls** [comment]
Note than `if (is_array($this->app->config['view']['paths'])) {` in https://github.com/laravel/framework/blob/5.8/src/Illuminate/Support/ServiceProvider.php#L97

fails in PHP 7.4 on Lumen fresh app. It needs to test `is_array($this->app->config['view'])` too.

I will send a PR to `driesvints:php-7.4` for this.

File: src/Illuminate/Support/ServiceProvider.php
Comment (GrahamCampbell) [inline src/Illuminate/Support/ServiceProvider.php]: Safer to do an isset on the full thing, with the paths access?

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/Illuminate/Support/ServiceProvider.php
Comment (kylekatarnls) [reply src/Illuminate/Support/ServiceProvider.php]: The isset is only on `['view']`, paths is not concerned.

+ See my comment above.

Comment (kylekatarnls) [reply src/Illuminate/Support/ServiceProvider.php]: IMHO, doing `isset($this->app->config['view'], $this->app->config['view']['paths'])` is overkill as you're not supposed to configure view without paths. However `isset($this->app->config['view'])` is required as view can be not configured (with a fresh install of Lumen for example). And in this case PHP 7.4 throws an error when running `null['paths']`

Comment (GrahamCampbell) [reply src/Illuminate/Support/ServiceProvider.php]: Just do `isset($this->app->config['view']['paths'])`.

Comment (kylekatarnls) [reply src/Illuminate/Support/ServiceProvider.php]: In which case `config['view']` could be set while `config['view']['paths']` would be null?

Comment (kylekatarnls) [reply src/Illuminate/Support/ServiceProvider.php]: OK.

Comment (kylekatarnls) [reply src/Illuminate/Support/ServiceProvider.php]: Done.

File: src/Illuminate/Support/Traits/Macroable.php
Comment (GrahamCampbell) [inline src/Illuminate/Support/Traits/Macroable.php]: Can we replace `call_user_func_array` with just `$macro(...$parameters)`?

Comment (driesvints) [reply src/Illuminate/Support/Traits/Macroable.php]: Done. Good suggestion 👍

**Comment by GrahamCampbell** [comment]
Those reflection errors are caused by the phpunit version being too low. PHPUnit hides itself from the stack traces,

**Comment by GrahamCampbell** [comment]
https://github.com/driesvints/framework/pull/3 seems to fix the builds. 🎉

**Comment by driesvints** [comment]
@GrahamCampbell ah sorry. Didn't see your comments. Was figuring it out myself in the meantime :')

**Comment by driesvints** [comment]
Build is finally passing! 🥳

I've had to bump some minimum versions of packages that implemented PHP 7.4 support:

- Symfony to v4.3.4: https://github.com/symfony/symfony/pull/32806
- PHPUnit to v8.3: https://github.com/sebastianbergmann/phpunit/pull/3765
- Mockery to v1.2.3: https://github.com/mockery/mockery/pull/980 
- EmailValidator to v2.1.10: https://github.com/egulias/EmailValidator/pull/203

Thanks @nikic and everyone else in this thread for your help!

**Comment by GrahamCampbell** [comment]
Part Parrot

**Comment by driesvints** [comment]
@GrahamCampbell can you dismiss your review or approve if everything is okay? :)

**Review by GrahamCampbell — APPROVED** [review]

**Comment by GrahamCampbell** [comment]
🚢

**Pull request merged**
